pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+// If split is set to false, it will write all result data into a single chunk in the series.
+// It writes a signal sample to the end of each chunk which preservers the last true sample of
+// the input data.
+// If it is called on already downsampled data, the input data must end at such a signaling
+// sample to recursively guarantee the semantics.
+func downsampleCounter(ser *series, data []sample, window int64, split bool) {
+	var (
+		n       = targetChunkSize(len(data))
+		nextT   = int64(-1)
+		lastT   int64
+		lastV   float64
+		counter float64
+	)
+	for i, s := range data {
+		if s.t > nextT {
+			if split && i > 0 && i%n == 0 {
+				// Add signaling sample that indicates what the true last value was
+				// before cutting a new chunk.
+				// For already downsampled series lastV is set to the last value that
+				// series itself added as a signaling sample.
+				ser.add(nextT, lastV)
+				ser.cut()
+			}
+			if nextT != -1 {
+				ser.add(nextT, counter)
+			}
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		if i == 0 {
+			// First sample sets the counter.
+			counter = s.v
+		} else if s.t > lastT {
+			// We ignored samples with no increasing timestamps. This handles signaling samples
+			// in already downsampled counter series.
+			// At breaking points (e.g. chunks) they add a sample with non-increased timestamp
+			// and the true last value, like the one we add ourselves in this function.
+			if s.v < lastV {
+				// Counter reset, correct the value.
+				counter += s.v
+			} else {
+				// Add delta with last value to the counter.
+				counter += s.v - lastV
+			}
+		}
+		lastV = s.v
+		lastT = s.t
+	}
+	ser.add(nextT, counter)
+	// Add signaling sample that indicates what the true last value was.
+	ser.add(nextT, lastV)
+	ser.cut()
+}
+
+func downsampleCounterCounter(ser *series, data []sample, chunked [][]sample, window int64) {
+	// For downsampling an already downsampled counter series, we have to ensure that produced
+	// chunks align with boundaries of input chunks.
+	// Otherwise the signal sample indiciating the last true sample value is not picked correctly.
+	n := targetChunkSize(len(data))
+	k := len(chunked)/(len(data)/n) + 1 // batch size
+
+	for len(chunked) > 0 {
+		j := k
+		if j > len(chunked) {
+			j = len(chunked)
+		}
+		length := 0
+		for _, p := range chunked[:j] {
+			length += len(p)
+		}
+		part := data[:length]
+		data = data[length:]
+		chunked = chunked[j:]
+
+		// Downsample the counter series spanning the chunk batch into a single chunk.
+		downsampleCounter(ser, part, window, false)
+	}
+}
+
+func downsampleCount(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	var count int
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, float64(count))
+			}
+			count = 0
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		count++
+	}
+	ser.add(nextT, float64(count))
+}
+
+func downsampleSum(ser *series, data []sample, window int64) {
```

Comment:
Ok, at that point I am voting for some aggregation struct that can downsample raw series in a single method (and let's say downsampled series in another (that may even reuse the former). These `downsampleXXX` are almost the same, maybe except the `counter`.  What do you think? This will really make it less prone to bugs and easier to read & maintain.

Reply:
I consider these functions pretty much write-once. So am not too concerned with duplication.

What I am worried about is that if we abstract the loop into something and than call some dynamically settable function(s), this will have significant performance impacts. Those loops are super hot – at the order of a few dozen billion runs each.

Reply:
hm, what I'd suggest rather is having single loop over `for _, s := range data {` that will fill `sum,max,min,count` in the same time instead of 4 separate functions. But there are some edge cases... like you can reuse sum,max,min functions but not count and counter.. I don't have clear alternative yet, so let's leave it.

count can be just sm

Reply:
Yea, counter is special but the others could indeed be done within one function. The only "problem" I suppose is that we just want a single one in the `downsampleAggr`.

I'll think about it again though and see if I can find something better.